server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -194,10 +196,19 @@ public ClusterState applyFailedShards(final ClusterState clusterState, final Lis
                         shardToFail.shardId(), shardToFail, failedShard);
                 }
                 int failedAllocations = failedShard.unassignedInfo() != null ? failedShard.unassignedInfo().getNumFailedAllocations() : 0;
+                final Set<String> failedNodeIds;
+                if (failedShard.unassignedInfo() != null &&
+                    failedShard.unassignedInfo().getReason() == UnassignedInfo.Reason.REALLOCATED_REPLICA) {
```

Comment:
Duplicating https://github.com/elastic/elasticsearch/pull/48265#discussion_r340525608:

Hmm. I do agree that if a shard is failed while started we should not count it as an allocation failure (and therefore we should consider it for a future no-op recovery). However I think we should attempt to track all allocation failures here, not just those resulting from `REALLOCATED_REPLICA`. Can we revert 0dd5cc8 except for the change to the `else` branch that should read `failedNodeIds = Collections.emptySet();`?